### DISCUSSION TITLE
Thoughts on an MJPEG effect?

### DISCUSSION THREAD

**Discussion by davepl** [PlummersSoftwareLLC/NightDriverStrip#533] [General]
Since we already have the heavy lifting done in terms of JPEG display anyway, I was thinking it might be cool to add an effect that can loop MJPEG files.  Except they'd have to go in the flash... so that raises two issues:

1) How much flash can we afford to spare, and how big are MJPEG files at 64x32?
2) Where can I find 64x32 versions of the dancing banana, the rainbow cat thing, and all the common anims?

**Comment by robertlipe** [comment]
SD cards can be considered infinitely large. They're only a few pins. Fewer pins == slower, but you could probably dupont one up to official Mesmerizer if you're not working with a bare board where 16MB Flash 8MB RAM cards are about $5.

Marc Merlin runs a few projects that are almost cousins of this one. The collection he runs on the HUB75's under his shirt at Burning Man seems to be:

https://github.com/marcmerlin/AnimatedGIFs

If I don't have the path quite right, it's in one of his other GH projects. He had a collection of a hundred or so that were sampled down to look good on 32x32 and 64x64 panels.

See also: https://marc.merlins.org/perso/arduino/post_2018-07-13_AnimatedGIFs-for-SmartMatrix-or-NeoMatrix-_Neopixel-WS2812B_-from-SDcard-or-SPIFFS_-on-Teensy-3_x_-ESP8266_-or-ESP32.html and https://marc.merlins.org/perso/arduino/post_2023-06-16_Party-LED-Outfit-Version-6_-Self-Contained-Power-System-on-LED-Panels.html as alternate entry points to his collection.


Related: there are flash-aware versions of FAT32 around the web for ESP32, but rumor has it that even those versions tend to eat cards pretty quickly under regular writes. (I've been working on additional filesystems in the background anyway, with LittleFS being my next stop.)

**Reply by davepl** [reply]
Looking at the stuff you linked, it's actually based on SmartMatrix's GIF support by Louis Beaudoin, so I need to look into what support we already have, maybe animated GIF is better than MJPEG, not sure.

We don't have an SD card by design, and won't be adding one to the mesmerizer, but if some other hardware platform has everything we need, we could look at it.  For now I'm pretty sure there aren't enough pins for debugging, HUB75, mic, IR, and an SD connection... so you'd lose debugging at least.

**Reply by mikejohnau** [reply]
I've downloaded a couple of the 64x64 GIFs and rendered them as MP4 video in Photoshop. 
GIF file size - 29kb
MP4 file size - 201kb.

A high cost for MP4 on a device with limited storage.

**Reply by mikejohnau** [reply]
https://github.com/PlummersSoftwareLLC/NightDriverStrip/assets/26905608/df039ea0-04af-4cb7-989d-e71d95c87a7a

**Comment by robertlipe** [comment]
It's not my specialty, but I intended to focus on Animated GIF when I moved
to this part of my Mesmerizer extension.

I can't answer #1. I'm far from my Mesmerizer.
I tried to help with #2.


TBC, Animated JPG, Animated GIF, and MP4/MPEG4 are three different things.

MP4/MEG4 is all about video quality; it's what you would use for a movie,
not a dancing banana.

My understanding is that:

MJPEG has the nice trait that every frame is a key frame. Its weakness is
that every frame is a key frame. It's like a flipbook animation with each
page/frame benign slightly different, but capable of existing on its own.
Like JPEG, they're good at photograph-y things.

Animated GIF is similar, but has some frame->frame compression; you can't
just seek into the middle of it as frame X+1 may depend on renderign frame
X.  Like GIF, they're good for low-res, high contrast things.  (The kinds
of things that look really good on HUB75s.)

MPEG4 and MP4 deserve a book of their own. Small file size is zero of the
reasons to use these. What works well for Avatar doesn't work well for
Dancing Bananas.


I actually know Marc. His collection is very much driven by his own desire
to run multiple panels like sandwich boards during Burning Man. He could
afford compute time up front to run shell scripts to convert anything to
anything else from his Linux computer to something that would run well on
the Pi's that replaced the Teensys he used to use. Looking good and fitting
on reasonable storage were goals.

Further, elsewhere in his GitHub collection, he runs a collection of a
SmartMatrix-like substance that's meant to unify many of the "render an
array of X*Y pixels into a grid" software APIs into one well enough to run
a common subset that's approximately analogous to our
include/effects/matrix/*.h effects. His goal in building this was both to
make software integration for another software/hardware hacker less painful
AND to collect effects libraries in pursuit of his Burning Man and clubbing
<https://marc.merlins.org/perso/arduino/post_2019-04-08_Clubbing_-EDM-Festival-and-Burning-Man-LED-Pants-and-Shirt-v4-on-ESP32-with-RGBPanels-and-SmartMatrix_GFX.html>
.

**Comment by davepl** [comment]
Here's a quick demo of what I came up with...

https://github.com/PlummersSoftwareLLC/NightDriverStrip/assets/3698967/e0922727-6985-4676-bfc9-24bbb141168e

**Reply by davepl** [reply]
And one more example...

https://github.com/PlummersSoftwareLLC/NightDriverStrip/assets/3698967/ea7d3583-45b5-4dc9-8f35-300825bf6315

**Comment by robertlipe** [comment]
Nice! But do the Bananas Dance?

**Reply by davepl** [reply]
Haven't found a good one designed to go on a black background yet!

**Comment by robertlipe** [comment]
More concisely, what input formats does this new reader of yours use?

Oooh, good point. White backgrounds on those panels is like staring at the
sun.

Peripherally related. (*Not really related*...) Ali's BF sale has free
shipping, 64x64 HUB75's for < $11.41 today, limit 1.

https://www.aliexpress.us/item/3256805834041887.html

They're P3's, so they're not awesome, but they're less than shipping
normally is on a 64x64....

**Comment by robertlipe** [comment]
There are only eight frames in
https://tenor.com/view/banana-dancing-cheering-gif-13570979

Would a transparent background render as black in whatever decoder you're
using?

This one looks like it's encoded with a transparent background, but it has
the problem of lots of colors that bananas aren't.
https://gifdb.com/images/high/dancing-color-changing-banana-uqgjzgre642ls20t.webp
or
https://gifdb.com/images/high/dancing-color-changing-banana-uqgjzgre642ls20t.gif
might be easier to work with. If it needs a border, we can fix that - the
question is about transparency; otherwise we have to pick a background
color.

If I were to make the background in those eight frames something else, what
color should the outlines (currently black) be?

For science, you know... and totally not be because I'm trapped at family
for holiday travels in a house full of people that sleep at normal times.

**Comment by heidepiek** [comment]
Brian lough had a dancing banana. 

https://github.com/PlummersSoftwareLLC/NightDriverStrip/assets/104028471/77592fd6-40c5-4dbf-978a-343bdaa6f24d

https://github.com/witnessmenow/LED-Matrix-Display-Examples

Harrie

**Comment by davepl** [comment]
It took a lot of shenanigans with exgif.com, but I made a nice 64x32 pacman chase that comes in at 35K...

**Comment by davepl** [comment]
Demo time!

https://github.com/PlummersSoftwareLLC/NightDriverStrip/assets/3698967/739e115c-1ccf-4a9c-950d-b14bbe1a02db

**Reply by davepl** [reply]
Arguably they should be blue when pacman is chasing them, but I lack the art skills to fix that :-)
